### PR TITLE
Filter tracks extrapolated to the muon system

### PR DESCRIPTION
**Pull request by CeliaFernandez**
#### PR description:

This PR modifies the track extrapolation to the muon chambers to reduce the time taken in muons1stStep sequence. In the current setup all tracks (passing soft pt and eta requirements) are extrapolated to all chambers that are geometrically compatible with the track parameters. The proposed changes allow to check if there are segments/hits in these chambers before initiating any extrapolation. If there are segments/hits, the track is extrapolated to the compatible chambers, if not, extrapolation is aborted.

Implementation was built on top of 13_0_0_pre3.

The time taken by muon1stStep was reduced from 4.5% to 3.8%, which translates into a reduction of total RecoMuon time from 7.0% to 6.1%. This measurement was done in lxplus with low precision, so checks are required.

**No changes are expected in any collection as this PR is not intended to modify any physics behavior but reduce the time taken by muon reconstruction.**

More info in this presentation: https://indico.cern.ch/event/1257332/

Some notes:
1. ME0 segments were loaded by default, so ME0 hits were also loaded as well. There is an ongoing migration from GEM → ME0 ( [#31424](https://github.com/cms-sw/cmssw/issues/31424)), should we take advantage here to remove them from TrackAssociator?
2. There is a merge conflict with changes introduced by other PR (https://github.com/cms-sw/cmssw/pull/40698), although the proposed changes should give an equivalent outcome. How can I proceed here?

#### PR validation:

This PR was tested with standard 11834.21 workflow. It passes the runTheMatrix tests with the exception of 2 where the connection of the input samples couldn't be readed.

#### If this PR is a backport please specify the original PR and why you need to backport that PR. If this PR will be backported please specify to which release cycle the backport is meant for:

This PR is not a backport.

### CODE REVIEW

File: TrackingTools/TrackAssociator/src/TrackAssociatorParameters.cc [hunk 1]
```diff
@@ -95,3 +106,44 @@ void TrackAssociatorParameters::loadParameters(const edm::ParameterSet& iConfig,
...
+  desc.add<edm::InputTag>("DTRecSegment4DCollectionLabel", edm::InputTag("dt4DSegments"));
+  desc.add<edm::InputTag>("EBRecHitCollectionLabel", edm::InputTag("ecalRecHit", "EcalRecHitsEB"));
+  desc.add<edm::InputTag>("EERecHitCollectionLabel", edm::InputTag("ecalRecHit", "EcalRecHitsEE"));
+  desc.add<edm::InputTag>("GEMSegmentCollectionLabel", edm::InputTag("gemSegments"));
+  desc.add<edm::InputTag>("HBHERecHitCollectionLabel", edm::InputTag("hbreco"));
+  desc.add<edm::InputTag>("HORecHitCollectionLabel", edm::InputTag("horeco"));
+  desc.add<edm::InputTag>("ME0SegmentCollectionLabel", edm::InputTag("me0Segments"));
+  desc.add<edm::InputTag>("RPCHitCollectionLabel", edm::InputTag("rpcRecHits"));
+  desc.add<edm::InputTag>("GEMHitCollectionLabel", edm::InputTag("gemRecHits"));
+  desc.add<edm::InputTag>("ME0HitCollectionLabel", edm::InputTag("me0RecHits"));
```
Comment (mmusich) [inline TrackingTools/TrackAssociator/src/TrackAssociatorParameters.cc:148]:
```suggestion
```

this should already been taken care of from https://github.com/cms-sw/cmssw/pull/40698

Comment (CeliaFernandez) [reply TrackingTools/TrackAssociator/src/TrackAssociatorParameters.cc:148]: Hi @mmusich ,

I also added some parameters to the fillDescriptions method that are necessary for this PR:
L124: preselectMuonTracks
L146: RPCHitCollectionLabel
L147: GEMHitCollectionLabel
L148: ME0HitCollectionLabel

Comment (mmusich) [reply TrackingTools/TrackAssociator/src/TrackAssociatorParameters.cc:148]: then add just those.

File: RecoEcal/EgammaClusterProducers/BuildFile.xml [hunk 2]
```diff
@@ -4,6 +4,8 @@
 <use name="DataFormats/EcalDigi"/>
 <use name="DataFormats/EcalRecHit"/>
 <use name="DataFormats/EgammaReco"/>
+<use name="DataFormats/RPCRecHit"/>
+<use name="DataFormats/GEMRecHit"/>
```
Comment (thomreis) [inline RecoEcal/EgammaClusterProducers/BuildFile.xml:8]: Why are these changes needed?

Comment (CeliaFernandez) [reply RecoEcal/EgammaClusterProducers/BuildFile.xml:8]: @thomreis  TrackAssociatorParameters are loaded in https://github.com/cms-sw/cmssw/blob/b0b0e732c03ad78d999359313705e18a767c6ddb/RecoEcal/EgammaClusterProducers/src/InterestingTrackEcalDetIdProducer.cc#L84 and there were some new parameters added to the TrackAssociatorParameters class. Without these lines I got a break segmentation fault when I was first implementing it.

**Comment by mmusich** [comment]
@cmsbuild, please test

**Comment by mmusich** [comment]
enable profiling

**Comment by mmusich** [comment]
> No changes are expected in any collection as this PR is not intended to modify any physics behavior but reduce the time taken by muon reconstruction.

Looks like this PR is causing genuine changes: https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_1_X_2023-04-12-1100+619298/56509/validateJR.html

**Comment by CeliaFernandez** [comment]
> > No changes are expected in any collection as this PR is not intended to modify any physics behavior but reduce the time taken by muon reconstruction.
> 
> Looks like this PR is causing genuine changes: https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_1_X_2023-04-12-1100+619298/56509/validateJR.html

Thanks @mmusich , I did not observe these changes before... Indeed the proposed changes are not fully transparent and this effects need to be understood.
I see that the number of muons remains the same but the segment information has changed for some of them. From a quick look to the matched segment size plots it seems that the variation is happening in muons where the matches (segments/hits) are always zero e.g.

- https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_1_X_2023-04-12-1100+619298/56509/validateJR/9.0_Higgs200ChargedTaus/all_RECO_step3/c_recoMuons_muons__RECO_obj_muMatches__segmentMatchesAT_size.png
- https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_1_X_2023-04-12-1100+619298/56509/validateJR/9.0_Higgs200ChargedTaus/all_RECO_step3/c_recoMuons_muons__RECO_obj_muMatches__gemMatchesAT_size.png
- https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_1_X_2023-04-12-1100+619298/56509/validateJR/9.0_Higgs200ChargedTaus/all_RECO_step3/c_recoMuons_muons__RECO_obj_muMatches__me0MatchesAT_size.png
- https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_1_X_2023-04-12-1100+619298/56509/validateJR/9.0_Higgs200ChargedTaus/all_RECO_step3/c_recoMuons_muons__RECO_obj_muMatches__rpcMatchesAT_size.png

And at the same time, the number of chambers seems to be "switching" some of the values to zero:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/baseLineComparisons/CMSSW_13_1_X_2023-04-12-1100+619298/56509/validateJR/9.0_Higgs200ChargedTaus/all_RECO_step3/c_recoMuons_muonsFromCosmics__RECO_obj_numberOfChambers.png

I suspect that the muons suffering these changes are those whose tracks are successfully extrapolated to chambers but none of the extrapolated chambers have segments. These were counted before when all the extrapolations were done. However here, where the extrapolations are done only if at least one of the chambers have segments, these are not counted. Since tracker muons have at least one segment (the implementation was done taking this into account) I suspect that the affected muons are global/standalone muons which do not make it to tracker muons.

I will do more studies to check is this is the case and try to fully understand what is happening...

**Comment by mandrenguyen** [comment]
@CeliaFernandez Any update here?  Would you say these differences are now expected?

**Comment by mandrenguyen** [comment]
+reconstruction
Changes are in line with expectations as described in https://github.com/cms-sw/cmssw/pull/41323#issuecomment-1582610727
Didn't occur to me at the ORP, but would be nice to squeeze this into 13_2_0_pre2

**Comment by rappoccio** [comment]
+1
